### PR TITLE
feat: Add more output to repair statistics

### LINKED ISSUE #246: add a JSON Schema for the JSON output
@khaes-kth @fermadeiral will meet to create first alpha version

### PR DESCRIPTION
**Pull request by slarse**
Fix #246 

This turned out to be more work than I thought. The granularity of the data listed in the schema is high, and it was non-trivial to get put it together.

Still working on it, but I'll be done tomorrow morning.

### CODE REVIEW

**Comment by slarse** [comment]
I suspect the failure is due to an overflow somewhere, but I can't find that I'm using an `int` anywhere ...

File: src/main/java/sorald/cli/Cli.java [hunk 1]
```diff
@@ -238,7 +290,7 @@ private RuleViolation parseRuleViolation(String violationSpecifier) {
             int startCol = Integer.parseInt(parts[3]);
             int endLine = Integer.parseInt(parts[4]);
             int endCol = Integer.parseInt(parts[5]);
-            return new SpecifiedViolation(key, fileName, startLine, endLine, startCol, endCol);
+            return new SpecifiedViolation(key, fileName, startLine, startCol, endLine, endCol);
```
Comment (slarse) [inline src/main/java/sorald/cli/Cli.java:293]: Targeted repair was apparently entirely broken. The stats collector tests failed miserably before I found this error.

**Comment by slarse** [comment]
@khaes-kth I think this is ready for you now. It turned out to be quite hard to give the highly granular statistics that were requested, but I think I've gotten the most crucial parts.

Here is sample output from a repair:

```json
{
    "repairs": [{
        "nbPerformedRepairs": 1,
        "ruleName": "XxeProcessing",
        "crashedRepairsLocations": [],
        "nbCrashedRepairs": 0,
        "ruleKey": "2755",
        "performedRepairsLocations": [{
            "endLine": 15,
            "endColumn": 42,
            "startColumn": 31,
            "startLine": 15,
            "filePath": "2755_XxeProcessing/InsecureXMLInputFactoryUsage.java",
            "violationSpecifier": "2755:2755_XxeProcessing/InsecureXMLInputFactoryUsage.java:15:31:15:42"
        }],
        "nbViolationsBefore": 7,
        "nbViolationsAfter": 6
    }],
    "executionInfo": {
        "soraldVersion": "LOCAL",
        "javaVersion": "11.0.8",
        "originalArgs": [
            "repair",
            "--original-files-path",
            "/tmp/junit6634076807002230511/project",
            "--stats-output-file",
            "/tmp/junit6634076807002230511/stats.json",
            "--file-output-strategy",
            "IN_PLACE",
            "--violation-specs",
            "2755:2755_XxeProcessing/InsecureXMLInputFactoryUsage.java:15:31:15:42"
        ]
    },
    "totalTimeMs": 987,
    "repairTimeMs": 132,
    "startTimeMs": 1607682619618,
    "crashes": [],
    "endTimeMs": 1607682620605,
    "parseTimeMs": 466
}
```

I have logically divided repairs into two different kinds: _performed_ and _crashed_. A performed repair is not necessarily successful, it just means that the repair exited without errors. A _crashed_ repair is a repair in which the processor attempting the repair suffered an unrecoverable error. If everything is as it should, then `nbViolationsAfter = nbViolationsBefore - nbPerformedRepairs`. **This must be verified for each stats file**, as there may be bugs in our processors that cause them to fail repairs _without_ crashing.

Note that I have not included detailed running times for individual rules, as that would entail quite a lot more work, and it seems rather pointless at this time as we're doing targeted repair of single violations anyway.

The naming doesn't quite match the schema in #246, as I found that there were better names to use. It also follows @monperrus suggestion in #284 to prefer "violation" over "warning".

**Comment by slarse** [comment]
@nharrand If you want to kick off a few experiments ASAP, you can just grab this branch. Usage won't change, and I suspect we still have bugs that need to be fixed before the actual experiments can take place.

**Comment by khaes-kth** [comment]
Thank you @slarse for adding the desired stats to the output.

I think it is now ready to be used by @nharrand for the experiment. WDYT?

**Comment by slarse** [comment]
@khaes-kth I think we will encounter a lot of crashes because of problems in Spoon, the sniper printer in particular. I'm working on it but I don't think we'll ever get it perfect. We might need something like #288 in order for the experiments to be reasonably successful.

Alternatively, we may just want to run in segment mode to begin with. Either way, I must add statistics collection to segment mode, so I'll get on that right away.

**Pull request merged**